File: rust/parallel_mergesort.rs
use std::env;
use std::fs::File;
use std::io::{BufRead, BufReader};
use std::time::Instant;
use rayon::prelude::*;

fn parallel_merge_sort(arr: &mut [i32]) {
    if arr.len() > 1 {
        parallel_merge_sort_recursive(arr);
    }
}

fn parallel_merge_sort_recursive(arr: &mut [i32]) {
    const SEQUENTIAL_THRESHOLD: usize = 10_000; // Switch to sequential for small arrays
    
    if arr.len() <= SEQUENTIAL_THRESHOLD {
        // Use sequential merge sort for small arrays
        arr.sort();
    } else if arr.len() > 1 {
        let mid = arr.len() / 2;
        
        // Split the array into two mutable parts
        let (left_half, right_half) = arr.split_at_mut(mid);
        
        // Use rayon to parallelize the recursive calls
        rayon::join(
            || parallel_merge_sort_recursive(left_half),
            || parallel_merge_sort_recursive(right_half),
        );
        
        // After sorting, merge the two halves using temporary storage
        let left_copy: Vec<i32> = left_half.to_vec();
        let right_copy: Vec<i32> = right_half.to_vec();
        merge_halves(&left_copy, &right_copy, arr);
    }
}

fn merge_halves(left: &[i32], right: &[i32], result: &mut [i32]) {
    let mut left_idx = 0;
    let mut right_idx = 0;
    let mut result_idx = 0;
    
    // Merge both halves into result
    while left_idx < left.len() && right_idx < right.len() {
        if left[left_idx] <= right[right_idx] {
            result[result_idx] = left[left_idx];
            left_idx += 1;
        } else {
            result[result_idx] = right[right_idx];
            right_idx += 1;
        }
        result_idx += 1;
    }
    
    // Copy remaining elements from left half
    while left_idx < left.len() {
        result[result_idx] = left[left_idx];
        left_idx += 1;
        result_idx += 1;
    }
    
    // Copy remaining elements from right half
    while right_idx < right.len() {
        result[result_idx] = right[right_idx];
        right_idx += 1;
        result_idx += 1;
    }
}

fn load_data(filename: &str) -> Result<Vec<i32>, Box<dyn std::error::Error>> {
    let file = File::open(filename)?;
    let reader = BufReader::new(file);
    let line = reader.lines().next().unwrap()?;
    
    let numbers: Result<Vec<i32>, _> = line
        .split(',')
        .map(|s| s.trim().parse::<i32>())
        .collect();
    
    Ok(numbers?)
}

fn is_sorted(arr: &[i32]) -> bool {
    arr.windows(2).all(|w| w[0] <= w[1])
}

fn is_prime(n: i32) -> bool {
    if n < 2 {
        return false;
    }
    if n == 2 {
        return true;
    }
    if n % 2 == 0 {
        return false;
    }

    let sqrt_n = (n as f64).sqrt() as i32;
    (3..=sqrt_n).step_by(2).all(|i| n % i != 0)
}

fn parallel_count_primes(numbers: &[i32]) -> usize {
    numbers.par_iter().filter(|&&num| is_prime(num)).count()
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args: Vec<String> = env::args().collect();
    let filename = args.get(1).map(|s| s.as_str()).unwrap_or("test_data.csv");

    println!("🦀 Rust Parallel Merge Sort + Prime Counting");
    println!("============================================");
    
    // Display system information
    let num_threads = rayon::current_num_threads();
    println!("Available CPU cores: {}", num_threads);
    println!("Parallel threshold: 10,000 elements");
    println!();

    // Load data
    println!("📊 Loading data...");
    let mut numbers = load_data(filename)?;
    println!("Loaded {} numbers", numbers.len());

    // Time the parallel sorting
    println!("🔄 Starting parallel merge sort...");
    let start_time = Instant::now();
    parallel_merge_sort(&mut numbers);
    let sort_end_time = Instant::now();

    let sort_time = sort_end_time.duration_since(start_time);
    println!("✅ Parallel merge sort completed in {:.4} seconds", sort_time.as_secs_f64());

    // Verify sorting is correct
    let sorted = is_sorted(&numbers);
    println!("🔍 Sorting verified: {}", sorted);

    // Count prime numbers in parallel
    println!("🔢 Starting parallel prime counting...");
    let prime_start_time = Instant::now();
    let prime_count = parallel_count_primes(&numbers);
    let prime_end_time = Instant::now();

    let prime_time = prime_end_time.duration_since(prime_start_time);
    let total_time = sort_time + prime_time;

    println!("✅ Parallel prime counting completed in {:.4} seconds", prime_time.as_secs_f64());
    println!("🎯 Found {} prime numbers", prime_count);
    println!("⏱️  Total execution time: {:.4} seconds", total_time.as_secs_f64());

    println!();
    println!("📈 Performance Details:");
    println!("- Rayon thread pool size: {}", num_threads);
    println!("- Sequential threshold: 10,000 elements");
    println!("- Sort time: {:.4}s", sort_time.as_secs_f64());
    println!("- Prime time: {:.4}s", prime_time.as_secs_f64());

    Ok(())
} 

File: go/parallel_mergesort.go
package main

import (
	"encoding/csv"
	"fmt"
	"os"
	"runtime"
	"strconv"
	"sync"
	"time"
)

// Threshold for switching to sequential sort for small arrays
const PARALLEL_THRESHOLD = 10000

func parallelMergeSort(arr []int, temp []int, left, right int, depth int) {
	if left < right {
		// Use sequential sort for small arrays or when we've used too many goroutines
		if right-left < PARALLEL_THRESHOLD || depth > 10 {
			sequentialMergeSort(arr, temp, left, right)
			return
		}

		mid := left + (right-left)/2
		
		// Use goroutines for parallel execution on larger arrays
		var wg sync.WaitGroup
		wg.Add(2)
		
		go func() {
			defer wg.Done()
			parallelMergeSort(arr, temp, left, mid, depth+1)
		}()
		
		go func() {
			defer wg.Done()
			parallelMergeSort(arr, temp, mid+1, right, depth+1)
		}()
		
		wg.Wait()
		merge(arr, temp, left, mid, right)
	}
}

func sequentialMergeSort(arr []int, temp []int, left, right int) {
	if left < right {
		mid := left + (right-left)/2
		
		sequentialMergeSort(arr, temp, left, mid)
		sequentialMergeSort(arr, temp, mid+1, right)
		merge(arr, temp, left, mid, right)
	}
}

func merge(arr []int, temp []int, left, mid, right int) {
	// Copy array section to temp buffer
	for i := left; i <= right; i++ {
		temp[i] = arr[i]
	}
	
	i, j, k := left, mid+1, left
	
	for i <= mid && j <= right {
		if temp[i] <= temp[j] {
			arr[k] = temp[i]
			i++
		} else {
			arr[k] = temp[j]
			j++
		}
		k++
	}
	
	// Add remaining elements from left side
	for i <= mid {
		arr[k] = temp[i]
		i++
		k++
	}
	
	// Add remaining elements from right side  
	for j <= right {
		arr[k] = temp[j]
		j++
		k++
	}
}

func loadData(filename string) ([]int, error) {
	file, err := os.Open(filename)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	reader := csv.NewReader(file)
	records, err := reader.ReadAll()
	if err != nil {
		return nil, err
	}

	if len(records) == 0 {
		return nil, fmt.Errorf("empty file")
	}

	numbers := make([]int, len(records[0]))
	for i, str := range records[0] {
		num, err := strconv.Atoi(str)
		if err != nil {
			return nil, err
		}
		numbers[i] = num
	}

	return numbers, nil
}

func isSorted(arr []int) bool {
	for i := 1; i < len(arr); i++ {
		if arr[i-1] > arr[i] {
			return false
		}
	}
	return true
}

func isPrime(n int) bool {
	if n < 2 {
		return false
	}
	if n == 2 {
		return true
	}
	if n%2 == 0 {
		return false
	}

	// Check odd divisors up to sqrt(n)
	for i := 3; i*i <= n; i += 2 {
		if n%i == 0 {
			return false
		}
	}
	return true
}

func countPrimesParallel(numbers []int) int {
	numWorkers := runtime.NumCPU()
	chunkSize := len(numbers) / numWorkers
	if chunkSize == 0 {
		chunkSize = 1
	}
	
	var wg sync.WaitGroup
	results := make([]int, numWorkers)
	
	for i := 0; i < numWorkers; i++ {
		wg.Add(1)
		go func(workerID int) {
			defer wg.Done()
			start := workerID * chunkSize
			end := start + chunkSize
			if workerID == numWorkers-1 {
				end = len(numbers) // Last worker takes remaining elements
			}
			
			count := 0
			for j := start; j < end; j++ {
				if isPrime(numbers[j]) {
					count++
				}
			}
			results[workerID] = count
		}(i)
	}
	
	wg.Wait()
	
	totalCount := 0
	for _, count := range results {
		totalCount += count
	}
	
	return totalCount
}

func main() {
	filename := "test_data.csv"
	if len(os.Args) > 1 {
		filename = os.Args[1]
	}

	// Get system info
	numCPU := runtime.NumCPU()
	runtime.GOMAXPROCS(numCPU)

	// Load data
	fmt.Println("Loading data...")
	numbers, err := loadData(filename)
	if err != nil {
		fmt.Printf("Error loading data: %v\n", err)
		return
	}
	fmt.Printf("Loaded %d numbers\n", len(numbers))
	fmt.Printf("Processors: %d\n", numCPU)
	fmt.Printf("GOMAXPROCS: %d\n", runtime.GOMAXPROCS(0))

	// Create temporary buffer for merge operations (single allocation)
	temp := make([]int, len(numbers))

	// Time the sorting
	fmt.Println("Starting parallel merge sort...")
	startTime := time.Now()
	parallelMergeSort(numbers, temp, 0, len(numbers)-1, 0)
	sortEndTime := time.Now()

	sortTime := sortEndTime.Sub(startTime).Seconds()
	fmt.Printf("Go parallel merge sort completed in %.4f seconds\n", sortTime)

	// Verify sorting is correct
	sorted := isSorted(numbers)
	fmt.Printf("Sorting verified: %t\n", sorted)

	// Count prime numbers in parallel
	fmt.Println("Counting prime numbers...")
	primeStartTime := time.Now()
	primeCount := countPrimesParallel(numbers)
	primeEndTime := time.Now()

	primeTime := primeEndTime.Sub(primeStartTime).Seconds()
	totalTime := sortTime + primeTime

	fmt.Printf("Prime counting completed in %.4f seconds\n", primeTime)
	fmt.Printf("Found %d prime numbers\n", primeCount)
	fmt.Printf("Total execution time: %.4f seconds\n", totalTime)
} 